src/utils/messaging-outbox/src/services/outbox_transactional_processor.rs

```diff
@@ -509,3 +530,39 @@ impl ProducerRelayJob {
 }
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug)]
+
+pub struct OutboxTransactionalProcessorMetrics {
+    pub failed_consumers: prometheus::IntGaugeVec,
+}
+
```

Comment:
This feels like `infra` layer to me. A trait could have been fine at service level, but a particular implementation that couples directly to `prometheus`, is not supposed to be here.

I understand `Prometheus` is de-facto a metrics standard, but there definitely are some alternatives, including OTEL collector itself via [OTEL SDK](https://docs.rs/opentelemetry/latest/opentelemetry/metrics/index.html).

Reply:
I spent a long time thinking about what is the best separation between our code and the metrics library. I started with the idea of isolation behind traits, as you suggested, but eventually changed my mind:

- `prometheus` is a very lightweight library - exposing it even in `domain` layer (like we already do with much larger DataFusion / Arrow) would have negligible impact on compile times / dependency bloat
- The metrics primitives (gauges, counters, histograms) are low level - e.g. an `IntGauge` is just a wrapper around atomic integer. These objects can be used in complete isolation from the collector part of the library, so objects like my `OutboxTransactionalProcessorMetrics` can be used directly in tests, we can assert on values of the metrics, without any need for mocking traits
- The chances that we will switch from Prometheus to something else are near-zero
- Metrics are not a part of crate-to-crate communications interface. Service that provides metrics will implement them themselves, test them, but other crates will not be ever injecting or reading them. So switching from one library to another will only impact individual services and registration - it will not have any ripple effects in APIs.

So while large part of me wants a complete separation between core domain and dependencies, I think in this case it will not be very practical. Just like it would likely be a waste of time wrapping `DataFrame`s into our own traits, I think wrapping primitives like counters and gauges will have more overhead than value.

I'm open to changing my mind on this, but suggest that we start with this simpler one and refactor towards traits later if there is a real need. If we decide on traits later - this should be a very local non-impactful change.